### ISSUE TITLE
Rule inlining

### ISSUE THREAD

**Issue by isidentical** [we-like-parsers/pegen#6]
On an example grammar like this, 
```
@subheader """\
import ast
import itertools

def parse_number(token):
    for converter in [int, float, complex]:
        try:
            return converter(token.string)
        except ValueError:
            continue
    raise ValueError(f"Can't convert: {{number!r}}")

def parse_string(tokens):
    return ''.join(token.string for token in itertools.from_iterable(tokens))
"""

start[ast.mod]: a=expression NEWLINE* ENDMARKER { ast.dump(ast.Expression(a)) }

expression[ast.expr] (memo):
    | primary

primary[ast.expr]:
    | atom

atom[ast.expr]:
    | 'True' { ast.Constant(True) }
    | 'False' { ast.Constant(False) }
    | 'None' { ast.Constant(None) }
    | '...' { ast.Constant(Ellipsis) }
    | t_name
    | t_number
    | &STRING strings

t_name[ast.Name]: NAME { ast.Name(name.string, ast.Load()) }
t_number[ast.Constant]: NUMBER { ast.Constant(parse_number(number)) }

strings[ast.Constant] (memo): strings=STRING+ { ast.Constant(parse_string(strings)) }
```

When we have different rule groups (like `primary`, `expression` etc) this stuff gets piled up. So instead of seeing `Expression(body=Constant(value=1.0))` as the output, we see `Expression(body=[[[Constant(value=1.0)]]])`. This can be solved by adding actions that returns itself, like;
```
expression[ast.expr] (memo):
    | primary { primary }

primary[ast.expr]:
    | atom { atom }

atom[ast.expr]:
    | 'True' { ast.Constant(True) }
    | 'False' { ast.Constant(False) }
    | 'None' { ast.Constant(None) }
    | '...' { ast.Constant(Ellipsis) }
    | t_name { t_name }
    | t_number { t_number }
    | &STRING strings { strings }
```

or having an operator which actually does this under the hood. An example is the inline operator in the [lark-parser](https://lark-parser.readthedocs.io/en/latest/tree_construction.html#shaping-the-tree). This is how it might look
```
start[ast.mod]: a=expression NEWLINE* ENDMARKER { ast.dump(ast.Expression(a)) }

?expression[ast.expr] (memo):
    | primary

?primary[ast.expr]:
    | atom

?atom[ast.expr]:
    | 'True' { ast.Constant(True) }
    | 'False' { ast.Constant(False) }
    | 'None' { ast.Constant(None) }
    | '...' { ast.Constant(Ellipsis) }
    | t_name
    | t_number
    | &STRING strings

?t_name[ast.Name]: NAME { ast.Name(name.string, ast.Load()) }
?t_number[ast.Constant]: NUMBER { ast.Constant(parse_number(number)) }

?strings[ast.Constant] (memo): strings=STRING+ { ast.Constant(parse_string(strings)) }
```

**Comment by pablogsal** [comment]
I will very wary of adding more operators, as the grammar is getting more and more cluttered and difficult to read. Many of the new operators are also exclusive of the PEG grammar so readers normally are not very familiar to them and this all adds to cognitive complexity.

**Comment by gvanrossum** [comment]
> the grammar is getting more and more cluttered and difficult to read

Indeed. There are already things that I had to look up last time I looked at some section of the grammar. :-(

**Comment by isidentical** [comment]
What if we propagate all the rules' results if they only have one condition by default instead of adding a new operator to do it explicitly? So instead of returning `[primary]` or `[atom]` here, automatically return primary or atom. 
```
    @memoize
    def expression(self) -> Optional[ast . expr]:
        # expression: primary
        mark = self.mark()
        cut = False
        if (
            (primary := self.primary())
        ):
            return [primary]
        self.reset(mark)
        if cut: return None
        return None

    @memoize
    def primary(self) -> Optional[ast . expr]:
        # primary: atom
        mark = self.mark()
        cut = False
        if (
            (atom := self.atom())
        ):
            return [atom]
        self.reset(mark)
        if cut: return None
        return None
```

**Comment by gvanrossum** [comment]
I have to admit I don't understand the use case. I'll leave it up to you
and Pablo.

**Comment by lysnikolaou** [comment]
Since the default action for alternatives with a single item is to return the item itself in the C generator, it makes sense for the Python generator to behave the same way. So let's go with that.

**Comment by isidentical** [comment]
Cool, can make a patch in the weekend

**Comment by MatthieuDartiailh** [comment]
This has been addressed as part of #11 so this could be closed, I believe.

**Comment by isidentical** [comment]
Ah, nice! Thanks
